java/sage/io/RemoteSageFile.java

```diff
@@ -0,0 +1,555 @@
+  }
+
+  /**
+   * Open a file remotely via the media server.
+   *
+   * @param hostname The hostname of the media server hosting this file.
+   * @param name The full path and file to be opened.
+   * @param readonly Set the channel to be read only.
+   * @param transcodeMode Set the transcode mode.
+   * @throws IOException If there is an I/O related error
+   */
+  public RemoteSageFile(String hostname, String name, boolean readonly, String transcodeMode) throws IOException
+  {
+    this.hostname = hostname;
+    remoteFilename = name;
+    this.transcodeMode = transcodeMode;
+    this.readonly = readonly;
+    this.forceActive = false;
+    connect();
+  }
+
+  /**
+   * Open a file remotely via the media server.
+   *
+   * @param hostname The hostname of the media server hosting this file.
+   * @param file The file to be opened.
+   * @param readonly Set the channel to be read only.
+   * @param forceActive Always assume the file is active. Use this with care.
+   * @throws IOException If there is an I/O related error
+   */
+  public RemoteSageFile(String hostname, File file, boolean readonly, boolean forceActive) throws IOException
+  {
+    this(hostname, file.getPath(), readonly, forceActive);
+  }
+
+  /**
+   * Open a file remotely via the media server.
+   *
+   * @param hostname The hostname of the media server hosting this file.
+   * @param name The full path and file to be opened.
+   * @param readonly Set the channel to be read only.
+   * @param forceActive Always assume the file is active. Use this with care.
+   * @throws IOException If there is an I/O related error
+   */
+  public RemoteSageFile(String hostname, String name, boolean readonly, boolean forceActive) throws IOException
+  {
+    this.hostname = hostname;
+    remoteFilename = name;
+    this.transcodeMode = null;
+    this.readonly = readonly;
+    this.forceActive = forceActive;
+    connect();
+  }
+
+  private void connect() throws IOException
+  {
+    if (socket != null)
+      return;
+
+    Socket newSocket = new Socket();
+    newSocket.connect(new InetSocketAddress(hostname, 7818));
+    newSocket.setSoTimeout(30000);
+    newSocket.setTcpNoDelay(true);
+    socket = newSocket;
+
+    outStream = new DataOutputStream(new BufferedOutputStream(socket.getOutputStream()));
+    inStream = new DataInputStream(new BufferedInputStream(socket.getInputStream()));
+
+    if (transcodeMode != null && transcodeMode.length() > 0)
+    {
+      outStream.write(("XCODE_SETUP " + transcodeMode + "\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      String response = Sage.readLineBytes(inStream);
+      if (!"OK".equals(response))
+        throw new java.io.IOException("Error with remote transcode setup for " + transcodeMode + " of: " + response);
+    }
+
+    if (readonly)
+      outStream.write("OPENW ".getBytes(Sage.BYTE_CHARSET));
+    else
+      outStream.write("WRITEOPENW ".getBytes(Sage.BYTE_CHARSET));
```

Comment:
WRITEOPENW takes 2 paramters, the first is the file path, and the second is the uploadID. That's an integer value that's retrieved from the server to prevent any other network service form arbitrarily opening a file and being able to write to it. I'm fine with you making that happen as part of this automatically. The way you get that on the client is calling SageTVConnection.requestUploadSpace(File, long) which specifies a file path and how big the upload is going to be. It uses the file size for freeing up space if needed in case that file path is in a controlled directory. However, apparently that takes zero size to clear the request...so you need to put some size in there. It also fails if the file already exists. (look at SageTVConnection.recvRequestUploadSpace which is the server end of the implementation). Maybe make a -1 file size a special case in there which doesn't free up any space, disregards checks for existing files and just gets a new upload ID.